File: Python/dfs_bfs/BJ2630.py
white = 0; blue = 0;

def cutPaper( row, col, size ):
    if chkCol( row, col, size ):
        if arr[row][col] == 0: white += 1
        else: blue += 1
        return
    
    newSize = size // 2;
    cutPaper(row, col, newSize)
    cutPaper(row, col+newSize, newSize)
    cutPaper(row+newSize, col, newSize)
    cutPaper(row+newSize, col+newSize, newSize)
    
def chkCol( row, col, size ):
    color = arr[row][col];
    
    for i in range( row, row+size ):
        for j in range( col, col+size ):
            if arr[i][j] != color: return False;
    return True;

n = int( input() )
arr = [ [0]*n for _ in range( n ) ]

cutPaper( 0, 0, n )

print( white ); print( blue )

File: src/Java/bfs_dfs/BJ2630.java
package bfs_dfs;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.StringTokenizer;

public class BJ2630 {
	private static int white = 0;
	private static int blue = 0;
	private static int[][] arr;
	
	private static boolean colChk( int row, int col, int size ) {
		int color = arr[row][col];
		
		for( int i=row; i<row+size; i++ ) {
			for( int j=col; j<col+size; j++ ) {
				if( arr[i][j] != color ) return false;
			}
		}
		
		return true;
	}

	private static void cutPaper( int row, int col, int size ) {
		if( colChk( row, col, size ) ) {
			if( arr[row][col] == 0 ) white++;
			else blue++;
			
			return;
		}
		
		int newSize = size / 2;
		cutPaper(row, col, newSize);
		cutPaper(row, col+newSize, newSize);
		cutPaper(row+newSize, col, newSize);
		cutPaper(row+newSize, col+newSize, newSize);
	}
	
	public static void main(String[] args) throws NumberFormatException, IOException {
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		int n = Integer.parseInt( br.readLine() );
		
		StringTokenizer st;
		arr = new int[n][n];
		for( int i=0; i<n; i++ ) {
			st = new StringTokenizer( br.readLine() );
			for( int j=0; j<n; j++ ) arr[i][j] = Integer.parseInt( st.nextToken() );
		}
		
		cutPaper(0, 0, n);
		
		System.out.println( white );
		System.out.println( blue );
		
		
		
	}
}